### ISSUE TITLE
Optional parameter is not cleared when navigating to same page with a different set of parameters

### ISSUE THREAD

**Issue by ThomasBleijendaal** [aspnet/AspNetCore#7419]
When navigating to a page which has two @page directives with one or two parameters, the optional one is not always cleared when navigating to this page when the optional parameter should have be set to its default.

In the [documentation](https://docs.microsoft.com/en-us/aspnet/core/razor-components/routing?view=aspnetcore-3.0#route-parameters) it is said to use two @page directives for support for optional parameters.

### To Reproduce:

Start with the sample app of Razor Components, and add a new page:

```
@page "/node/{Id:int}"
@page "/node/{ParentId:int}/{Id:int}"

@functions {
    [Parameter]
    private int? ParentId { get; set; } = null;

    [Parameter]
    private int Id { get; set; }
}
```

update NavMenu.cshtml to include:

```
<NavLink href="node/1">First click me</NavLink>
<NavLink href="node/2/3">Then click me</NavLink>
<NavLink href="node/4">And finally click me</NavLink>
```

Start the app.

Navigating to node/1 works: `ParentId` is null and `Id` is 1.
Navigating to node/2/3 works: `ParentId` is 2 and `Id` is 3.
Navigating to node/4 works but gives: `ParentId` *stays* 2, and `Id` is 4.

### Expected behavior

When navigating to node/4 I expect `ParentId` to be null again and `Id` to be 4. `ParentId` is a Parameter, and should be set to null when the route does not provide the information.

### Additional context

```.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview-010184
 Commit:    c57bde4593

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100-preview-010184\

Host (useful for support):
  Version: 3.0.0-preview-27324-5
  Commit:  63a01b08e5

.NET Core SDKs installed:
  1.0.0-preview2-003121 [C:\Program Files\dotnet\sdk]
  1.0.0-preview2-003131 [C:\Program Files\dotnet\sdk]
  1.0.0 [C:\Program Files\dotnet\sdk]
  1.0.4 [C:\Program Files\dotnet\sdk]
  1.1.0 [C:\Program Files\dotnet\sdk]
  2.0.0 [C:\Program Files\dotnet\sdk]
  2.0.3 [C:\Program Files\dotnet\sdk]
  2.1.2 [C:\Program Files\dotnet\sdk]
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.100 [C:\Program Files\dotnet\sdk]
  2.1.200 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.300-preview1-008174 [C:\Program Files\dotnet\sdk]
  2.1.300-rc1-008673 [C:\Program Files\dotnet\sdk]
  2.1.300 [C:\Program Files\dotnet\sdk]
  2.1.302 [C:\Program Files\dotnet\sdk]
  2.1.400 [C:\Program Files\dotnet\sdk]
  2.1.401 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.503 [C:\Program Files\dotnet\sdk]
  2.1.600-preview-009472 [C:\Program Files\dotnet\sdk]
  3.0.100-preview-010184 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0-preview1-final [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.0-rc1-final [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0-preview1-final [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.0-rc1-final [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview-19075-0444 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-preview1-26216-03 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-rc1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3-servicing-26724-03 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview-27324-5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview-27325-3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]```

**Comment by Andrzej-W** [comment]
This is very old bug and it was tracked somewhere. It was fixed but then reverted. See this #6864. Hopefully both problems will be solved in preview 3.

**Comment by mkArtakMSFT** [comment]
@SteveSandersonMS, is this fixed already?

**Comment by SteveSandersonMS** [comment]
@mkArtakMSFT We fixed it in 0.7.0, but then we had to revert the fix just prior to the 0.7.0 release because it caused a bad regression elsewhere. We don't yet have a design for what the replacement fix should be. It should remain on the backlog, but it's not at the top of the priority list in my opinion.

**Comment by mkArtakMSFT** [comment]
Thanks @SteveSandersonMS. Can you please dupe this one with that one you reference and close this issue?

**Comment by mkArtakMSFT** [comment]
@SteveSandersonMS, I assume you meant `0.8.0` there, did you?

**Comment by SteveSandersonMS** [comment]
Yes, 0.8.0. Thanks!

The change where we originally fixed this was #4797, but then we reverted it in #6931.

I've searched and can't find another issue on the backlog tracking making a further long-term fix, so perhaps this issue can be that one for us now :)

Putting on backlog.

**Comment by SteveSandersonMS** [comment]
Plan following offline discussion:

 * Make `Router` figure out the union of all the route parameter names across all the route templates for each component. Then when supplying parameters, supply `null` for any not in the current route.
 * Check that components can receive `null` for value-type parameters and they get treated as `default(T)`.

**Comment by RemiBou** [comment]
@SteveSandersonMS I might be mistaken, but this won't take into account properties default value and will erase them right (OP problem) ?

**Comment by SteveSandersonMS** [comment]
It will only erase parameters that match route tokens, not others.

**Comment by Paul-Schroeder** [comment]
In case anyone else stumbles upon this, I solved an issue where I thought the page wouldn't reload data based upon the parameters in the URL / Querystring changing.  What worked for me was overriding the "OnParametersSetAsync()" event handler (instead of code I had in the "OnAfterRenderAsync(bool firstRender)" event handler.  For example:

```
private readonly object _synclock = new object();

protected override async Task OnParametersSetAsync()
{
	IList<Model.MyModelClass> myModelClasses = await MyService.GetByStartUtcDateAsync(MinStartUtcDate);
	lock (_synclock)
	{
	ClearMyCollections();
	...ReloadMyCollections...
	}
}
```	

Note, I added the locking code because I was originally observing some inconsistent behavior where sometimes the collections I was loading contained up to three times the number of expected data records.  The OnParametersSetAsync() method seems to fire multiple times and I figured it was a threading issue; haven't dug too deep there.
